presto-orc/src/main/java/com/facebook/presto/orc/QualifyingSet.java

```diff
@@ -35,60 +38,51 @@
     private int[] inputNumbers;
     private ErrorSet errorSet;
 
-    static volatile int[] wholeRowGroup;
-    static volatile int[] allZeros;
+    private static volatile int[] wholeRowGroup;
+    private static volatile int[] allZeros;
     private int[] ownedPositions;
     private int[] ownedInputNumbers;
-    private QualifyingSet parent;
 
     static {
-        wholeRowGroup = new int[10000];
-        allZeros = new int[10000];
+        wholeRowGroup = new int[MAX_SIZE];
+        allZeros = new int[MAX_SIZE];
         Arrays.fill(allZeros, 0);
-        for (int i = 0; i < 10000; i++) {
+        for (int i = 0; i < MAX_SIZE; i++) {
             wholeRowGroup[i] = i;
         }
     }
 
-    public void setRange(int begin, int end)
+    public void setRange(int end)
     {
+        checkArgument(end >= 0, "end must not be negative");
         this.end = end;
         int[] zeros = allZeros;
-        if (zeros.length < end - begin) {
-            int[] newZeros = new int[end - begin];
+        if (zeros.length < end) {
+            int[] newZeros = new int[end];
             Arrays.fill(newZeros, 0);
             allZeros = newZeros;
```

Comment:
This maybe a race condition. If there are two threads, both read the allZero's length, and try to expand it, the allZeros may not be set to the larger of the two sizes. Making allZeros volatile is not enough, we probably need to use synchronized.